src/renderer/shadow.frag

```diff
@@ -0,0 +1,98 @@
+precision highp float;
+
+varying vec4 trapBounds;  // [x1, lineY1, x2, lineY2]
+varying vec4 trapBottom;  // [x1, bottomY1, x2, bottomY2]  
+varying vec2 worldPos;    // World position
+
+uniform int gradientCount;
+uniform sampler2D gradientTexture;  // Texture containing gradient data
+uniform vec4 fallbackColor;
+
+vec4 interpolateGradient(float position) {
+    if (gradientCount <= 0) {
+        return fallbackColor;
+    }
+    if (gradientCount == 1) {
+        return texture2D(gradientTexture, vec2(0.5 / float(gradientCount * 2), 0.5));
+    }
+
+    float textureWidth = float(gradientCount * 2);
+    float firstStop = texture2D(gradientTexture, vec2(0.5 / textureWidth, 0.5)).r;
+    float lastStop = texture2D(gradientTexture, vec2((float((gradientCount - 1) * 2) + 0.5) / textureWidth, 0.5)).r;
+    
+    // Scale position to fit the gradient range
+    if (lastStop > firstStop) {
```

Comment:
You can use a `mix` to get rid of this if, though if you make firstStop and lastStop uniforms it should get compiled out and be performant regardless